### DISCUSSION TITLE
Mirror (Model)

### DISCUSSION THREAD

**Discussion by jgomez720** [KittyCAD/modeling-app#508] [New Feature Requests]
### Description
A tool that allows the user to mirror an entire body across a plane.

### Functionality
1. The user selects the _Mirror_ button.
2. The user selects the body(ies) that they want to mirror.
3. The user selects the plane they want to mirror over.
4. Confirm.

### Supporting Docs/Images
![Screenshot 2023-11-09 at 2 07 39 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/da954ca9-d7aa-4cce-933a-ca234f3dc96f)
![Screenshot 2023-11-09 at 2 07 53 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/9647ea02-937b-44d8-b3d0-2f3c1aeb249e)
![Screenshot 2023-11-09 at 2 08 05 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/6c9f3405-b04d-42a7-b088-6678832b36b6)

## PseudoKCL

```
const part001 = startSketchOn('XY')
  |> startProfileAt([-12, 12], %)
  |> line([6, 0], %)
  |> line([0, -18], %)
  |> line([-6, 0], %)
  |> close(%)
  |> extrude(10, %)

const part002 = mirror({
  bodies: [part001], //option to have more here 
  mirror_plane: 'YZ', 
})
```
This is as far as I can get in current KCL.
![Screenshot 2023-11-09 at 2 20 08 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/0bca6c88-c685-48b7-93e8-69322b8e4ac1)

**What the result should look like:**
![Screenshot 2023-11-09 at 2 21 49 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/9c6ee6a2-12ea-4b80-88fc-bb6667794c6d)

**Comment by adamchalmers** [comment]
Isn't this just a pattern, where `numReplicas = 2`? We could write this KCL `mirror` function as a wrapper around `patternTransform` that creates another replica, translates it by a certain distance (2 * distance to the origin) and a certain direction (towards the origin).

This could be built purely in the UI/KCL with no engine support necessary!

**Comment by jgomez720** [comment]
It's more complex than that. Users are expecting to input a plane as a parameter. I gave an easy example above, but this can get a lot more complex. This should be able to mirror a geometry over _any_ plane. Meaning, something like this:

<img width="760" alt="Screenshot 2024-09-04 at 12 22 38 PM" src="https://github.com/user-attachments/assets/6cb417fb-9429-4d4f-85c3-61e2555bd57c">
<img width="809" alt="Screenshot 2024-09-04 at 12 22 47 PM" src="https://github.com/user-attachments/assets/8918e026-f99c-44d5-a17c-21e130b67a85">
